### ISSUE TITLE
Type compatibility/can assign

### ISSUE THREAD

**Issue by carloszimm** [cyberuni/type-plus#349]
Forgive me if I have understood wrong (the functionality) but the following code shouldn't work without compiler problems?
```typescript
assertType.isTrue(true as CanAssign<any, number>);
```
If the order is inverted, it works fine. Also, the following code apparently compiles:
```typescript
let a: any = 4;
let b: number = a;

let d: number = 3;
let e: any = d;
```
StackBlitz print:
![image](https://github.com/unional/type-plus/assets/4553211/539f76d3-9ecf-4ef7-9e97-169403d3ec7e)

Thank you :)

**Comment by unional** [comment]
Hi, yes, you are technically correct. TypeScript has some weird behavior around `any`.

Besides meaning any type, or union of all types, or the superset of all types, `any` is also used as a "type checking silencer", allowing user to tell the compiler that "hey, don't worry about the type of this".

That's why you can do `const x: number = 'a' as any`.

The `CanAssign<A, B>` type was checking more like on the set theory, which `any` cannot assign to `number`.

In fact, the `CanAssign<any, number>` should really return `boolean`, instead of `false`, as "some branch in the union (of `any`) can be assigned to `number`, while some cannot".

This type will be updated in the near future, but it will be a breaking change.

**Comment by unional** [comment]
Yes. The `IsAny` check can be applied. Many of the newer types such as `IsEqual` is doing that kind of checks.

`CanAssign` can be improved with that too.

For the inside a class/object, yes. That's the original purpose of the type and that is handled differently.

I'm working on a new way to handle branching to maximize flexibility. `CanAsssign` can use some updates when I get to it.

btw, if you want to check for `CanAssign` between two types, you can do `testType.canAssign<A,B>(true | false)`.

**Comment by carloszimm** [comment]
I wasn't aware that `testType` had a `canAssign` (the documentation apparently didn't mention it yet). Anyway, is it supposed to behave similarly to `assertType` (with the compiler complaining about the wrong type) or it should return only a boolean? If the later, why the following snippet is complaining? :(
![image](https://github.com/unional/type-plus/assets/4553211/ac5fa484-0feb-494b-b9f1-28537b996276)

Btw, I suppose `testType` is one of the types you mentioned that is making those checks, as it works correctly for the following test:
```javascript
console.log(testType.canAssign<number, any>(true)); //--> true
console.log(testType.canAssign<any, number>(true)); //--> true
```

**Comment by unional** [comment]
> I wasn't aware that `testType` had a `canAssign` (the documentation apparently didn't mention it yet).

Um, thanks for pointing it out. I thought I have the doc updated. Seems like I have missed it.

It is in [the docs under the `/testing` folder](https://github.com/unional/type-plus/blob/main/type-plus/ts/testing/readme.md#testtype),
and it has JSDocs and unit tests describing each function.

It is mainly for used during testing. The return value is merely a container of the type under inspection, as all the check are done with pure type anyway (the same for `assertType.isTrue(true ...)`, the return value can only be `true` 😄 )

For runtime check, currently the `canAssign<T>(): (subject: S) => CanAssign<S, T>` is the one you can use.
